### PR TITLE
Use selected subscription to determine azure credential

### PR DESCRIPTION
**Pull request by bishal-pdMSFT**
Currently azure session was hard-coded to use first session for credentials. However, in case of MSA user, there will be more than one azure sessions available and we need to use the appropriate one. This can only be detected if user does selection for subscription before we make any Azure/AzDevOps rest call. **Hence, I have moved Azure details to be asked prior to AzDevOps detail.**
I have tested with my MSA user for GitHub repo. Let me know if I also need to test for Azure Repos.

Also, added userType in telemetry to differentiate debugging failures.

### CODE REVIEW

File: src/configure/configure.ts [hunk 1]
```diff
@@ -432,16 +432,18 @@ class PipelineConfigurer {
 
     private async getAzureResourceDetails(): Promise<void> {
         // show available subscriptions and get the chosen one
-        let subscriptionList = extensionVariables.azureAccountExtensionApi.filters.map((subscriptionObject) => {
+        let subscriptionList = extensionVariables.azureAccountExtensionApi.subscriptions.map((subscriptionObject) => {
```
Comment (dikhakha) [inline src/configure/configure.ts]: I think using subscriptions instead of filters will list all the subscriptions even though user might have filtered out some subscriptions to be used. Do we want that ?

**Review by dikhakha — APPROVED** [review]
Left one comment

File: src/configure/configure.ts [hunk 2]
```diff
@@ -122,11 +120,11 @@ class PipelineConfigurer {
             this.inputs.githubPATToken = await this.getGitHubPATToken();
         }
 
-        await this.getAzureDevOpsDetails();
-
         if (!this.inputs.targetResource.resource) {
             await this.getAzureResourceDetails();
         }
+
+        await this.getAzureDevOpsDetails();
```
Comment (hiyadav) [inline src/configure/configure.ts:127]: will this not change the order in which inputs are asked from user ?

File: src/configure/configure.ts [hunk 3]
```diff
@@ -432,16 +432,18 @@ class PipelineConfigurer {
 
     private async getAzureResourceDetails(): Promise<void> {
         // show available subscriptions and get the chosen one
-        let subscriptionList = extensionVariables.azureAccountExtensionApi.filters.map((subscriptionObject) => {
+        let subscriptionList = extensionVariables.azureAccountExtensionApi.subscriptions.map((subscriptionObject) => {
             return <QuickPickItemWithData>{
-                label: <string>subscriptionObject.subscription.displayName,
+                label: `${<string>subscriptionObject.subscription.displayName} (${<string>subscriptionObject.subscription.subscriptionId})`,
```
Comment (hiyadav) [inline src/configure/configure.ts]: Why do we want to show subscription id ?

File: src/configure/configure.ts (hunk 1 above)
Comment (hiyadav) [inline src/configure/configure.ts]: should this remain filters ? or do we want to show all accessible subscriptions independent of Azure-account extension settings ?

File: src/configure/configure.ts [hunk 4]
```diff
@@ -432,16 +432,18 @@ class PipelineConfigurer {
...
                 data: subscriptionObject
             };
         });
         let selectedSubscription: QuickPickItemWithData = await this.controlProvider.showQuickPick(constants.SelectSubscription, subscriptionList, { placeHolder: Messages.selectSubscription });
         this.inputs.targetResource.subscriptionId = selectedSubscription.data.subscription.subscriptionId;
+        this.inputs.azureSession = getSubscriptionSession(this.inputs.targetResource.subscriptionId).session;
```
Comment (hiyadav) [inline src/configure/configure.ts]: Rather than initializing it at multiple places should we do it as part of some initialization.

File: src/configure/helper/azureSessionHelper.ts [hunk 5]
```diff
@@ -0,0 +1,14 @@
+import { extensionVariables, SubscriptionSession } from "../model/models";
+
+export function getSubscriptionSession(subscriptionId: string): SubscriptionSession {
```
Comment (hiyadav) [inline src/configure/helper/azureSessionHelper.ts]: should this return subscription.session, which corresponds to its name.

File: src/configure/helper/telemetryHelper.ts [hunk 6]
```diff
@@ -18,6 +18,9 @@ class TelemetryHelper {
         this.journeyId = uuid();
         this.command = command;
         this.setTelemetry(TelemetryKeys.JourneyId, this.journeyId);
+        if (process.env.DEBUG_MODE === "true") {
+            this.setTelemetry(TelemetryKeys.UserType, "test");
+        }
```
Comment (hiyadav) [inline src/configure/helper/telemetryHelper.ts]: this is not required, the sdk already takes care of this by changing the sessionId in the propertybag to this value: 'someValue'

File: src/configure/resources/telemetryKeys.ts [hunk 7]
```diff
@@ -1,5 +1,6 @@
 export class TelemetryKeys {
     public static CurrentUserInput: string = 'currentUserInput';
+    public static UserType: string = "userType";
```
Comment (hiyadav) [inline src/configure/resources/telemetryKeys.ts]: This will also not be required

File: src/extension.ts [hunk 8]
```diff
@@ -25,11 +25,11 @@ export async function activate(context: vscode.ExtensionContext) {
     await callWithTelemetryAndErrorHandling('azurePipelines.activate', async (activateContext: IActionContext) => {
         activateContext.telemetry.properties.isActivationEvent = 'true';
         telemetryHelper.initialize(activateContext, 'activate');
-        telemetryHelper.setTelemetry('configurePipelineEnabled', `${configurePipelineEnabled}`);
         await telemetryHelper.executeFunctionWithTimeTelemetry(
             async () => {
                 await activateYmlContributor(context);
                 if (configurePipelineEnabled) {
+                    telemetryHelper.setTelemetry('configurePipelineEnabled', `${configurePipelineEnabled}`);
```
Comment (hiyadav) [inline src/extension.ts]: the value of const `configurePipelineEnabled` is true or false, we can always log this irrespective of it being enabled or not. 
Our query will be like properties.configurePipelineEnabled == 'true'. 

We can revert this change

**Review by hiyadav — CHANGES_REQUESTED** [review]

File: src/configure/configure.ts (hunk 2 above)
Comment (bishal-pdMSFT) [reply src/configure/configure.ts:127]: Yes, but this is inevitable. We cannot get right credential to make REST calls until we let user choose a subscription.

File: src/configure/configure.ts (hunk 1 above)
Comment (bishal-pdMSFT) [reply src/configure/configure.ts]: I will stick with filters for now. User has decided to now work with an extension and we should honor that.

Comment (bishal-pdMSFT) [reply src/configure/configure.ts]: I was not aware of difference between filters and subscriptions. Will use filters.

File: src/configure/configure.ts (hunk 2 above)
Comment (hiyadav) [reply src/configure/configure.ts:127]: Okay, this makes sense.
Just for my knowledge, do we get multiple sessions on single login

Comment (hiyadav) [reply src/configure/configure.ts:127]: Got it from the description

File: src/configure/configure.ts (hunk 4 above)
Comment (bishal-pdMSFT) [reply src/configure/configure.ts]: I cannot find a good way to put these two lines in a separate method - it looks too forced. Hence I will leave it this way only.

File: src/configure/configure.ts (hunk 3 above)
Comment (bishal-pdMSFT) [reply src/configure/configure.ts]: Two subscriptions can have same name. When I was testing there were two MSDN subscription with exact same name - becomes very confusing.

File: src/configure/helper/telemetryHelper.ts (hunk 6 above)
Comment (hiyadav) [reply src/configure/helper/telemetryHelper.ts]: image

File: src/configure/configure.ts (hunk 3 above)
Comment (hiyadav) [reply src/configure/configure.ts]: In that case we can add this to the either description or details of the QuickPickItem, that display's it in less bold manner

**Review by hiyadav — APPROVED** [review]

**Comment by hiyadav** [comment]
Approved as major concern about inputs sequence change is resolved.

File: src/configure/helper/azureSessionHelper.ts (hunk 5 above)
Comment (bishal-pdMSFT) [reply src/configure/helper/azureSessionHelper.ts]: you are right, I can directly return .session. When I started I thought of using both session and subscription and hence was returning SubscriptionSession. But it is no longer required. Will return session directly and remove SubcriptionSession definition from model as well

File: src/configure/helper/telemetryHelper.ts (hunk 6 above)
Comment (bishal-pdMSFT) [reply src/configure/helper/telemetryHelper.ts]: Got it. Thanks. Removed.

File: src/configure/resources/telemetryKeys.ts (hunk 7 above)
Comment (bishal-pdMSFT) [reply src/configure/resources/telemetryKeys.ts]: removed

File: src/extension.ts (hunk 8 above)
Comment (bishal-pdMSFT) [reply src/extension.ts]: Got it. Reverted.

File: src/configure/configure.ts (hunk 3 above)
Comment (bishal-pdMSFT) [reply src/configure/configure.ts]: Changed to description
